examples/tool_tests/gowin.c

```diff
@@ -1,4 +1,6 @@
-#pragma PART "GW2AR-18C GW2AR-LV18QN88PC8/I7"
+// speed grade C8 is equivalent to I7 in this context, and resolves to the expected C8/I7
+// for gw_sh not to complain.
+#pragma PART GW2AR-LV18QN88PC8:C
 #pragma MAIN_MHZ my_pipeline 100.0
 // TEMP NOT UNTIL DONE DEV #include "examples/pipeline.c"
```

Comment:
Its looking great - want to try the bigger test case and ill get merged soon? uncomment this line and get rid of rest of file

Reply:
If missing IEEE floating point libraries in vhdl 2008 you might hit bump... will see :crossed_fingers:

Reply:
Trying it now, the report seems to have broken somehow generating the report for BIN_OP_PLUS_int31_t_int31_t - currently checking what happened, as the report _was_ generated 🤔

Reply:
I'll need to check that deeper, in fact it's failing somewhere else - UNARY_OP_NOT_uint25_t is generating a netlist that results in no timing info when analyzed by Gowin's tools.

Reply:
Hmm Interesting
Wonder if it is optimizing the inverter out of the path some how 

Does the gowin tool have any kind of retiming enabled by default?

Iirc I had to not use yosys retiming since it will retime and move logic to outside of register to register timing paths, and this the path won't be timed at all, and will declare met timing / no paths to time

Reply:
Or even if somehow gowin devices can integrate an inverter into their Flip flops and then declare not timing path?

Reply:
I think the way that it's doing it is your second hypothesis - even with retiming disabled (found an option that's not referred to by the Gowin Software Guide, `set_option -retiming 0` - hopefully it's working correctly) it outputs the same report with no timing info. The Gowin Primitives Guide (https://www.gowinsemi.com/upload/database_doc/1880/document/658e8288b2e42.pdf) does list INV as a mode for the LUTs/FFs, and the report mentions these in the summary.

Resource Usage Summary
Resource | Usage
-- | --
I/O Port | 51
I/O Buf | 51
&nbsp; IBUF | 26
&nbsp; OBUF | 25
Register | 50
&nbsp; DFF | 50
INV | 25
&nbsp; INV | 25

Resource Utilization Summary
Resource | Usage | Utilization
-- | -- | --
Logic | 25(25 LUT, 0 ALU) / 20736 | <1%
Register | 50 / 15750 | <1%
--Register as Latch | 0 / 15750 | 0%
--Register as FF | 50 / 15750 | <1%
BSRAM | 0 / 46 | 0%

Reply:
nice work finding retiming option - might actually be beneficial to turn back on once we know isnt root of this issue - is nice if it works correctly

Hmm the way it says it uses 25 LUTs made me think it wouldnt be the 'integrated into FF' thing 

The design as written is the 25 input FFs, 25 LUTs for inversion, 25 output FFs ... and somehow no timing paths for that huh?...

Hm... maybe hard code inversion as free/zero delay on gowin and see what happens...

Reply:
Re: wanting to hack in gowin + inverter = zero delay = dont trying to synthesize for a timing path 

The check for that condition is here
https://github.com/JulianKemmerer/PipelineC/blob/bd55aa790e761ddb15051247dcab172189d02ccd/src/SYN.py#L4589

And the 'lookup' of 'is this function zero delay' is here
https://github.com/JulianKemmerer/PipelineC/blob/bd55aa790e761ddb15051247dcab172189d02ccd/src/SYN.py#L4360

So could get to hackin .startswith(UNARY_OP_NOT_ etc kind of stuff :shrug:

Would obviously fall over in a design of just inverters... :grimacing:

Reply:
Maybe another option is trying to find some kind of HDL attribute like KEEP DONT TOUCH etc things that could disable merging inv into FFs (and can only turn that on when doing these single inverter path synthesis runs)

Reply:
^ to be clear the ~standard KEEP and DONT_TOUCH and iirc SYN_KEEP ? are generated on the input and output FFs already - but maybe need something gowin specific?

Reply:
According to their synthesis guide, they should support `syn_keep` - they also have `syn_preserve` and `syn_noprune`, which made no effect on the timings whatsoever. Can't rule out a bug in their synthesis timing calculations 😅

Reply:
Shorting the timing path synthesis did the trick, and the process went the whole way through 🎉

Reply:
^ Did it produce a floating point adder pipeline that met the ~100MHz timing goal on gowin? how many clock cycles/ pipeline stages did it take? (should be in last few lines of output)

im excited to try and collect some data - gotta install these gowin tools so I dont bother you as much :sweat_smile:

Reply:
```
my_pipeline Clock Goal: 100.00 (MHz) Current: 106.36 (MHz)(9.40 ns) 6 clks
Met timing...
================== Writing Results of Throughput Sweep ================================
Output VHDL files: /mnt/c/Users/tiago/Projects/PipelineC/pipelinec_output_gowin.c_1925/vhdl_files.txt
Done.
```

Reply:
Heck yeah :sunglasses: